lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,612 @@
+		return (err);
+	}
+
+	sa_bulk_attr_t sa_attrs[UZFS_END];
+	int cnt = 0;
+	libuzfs_add_bulk_attr(dhp, sa_attrs, &cnt, attr);
+	err = sa_bulk_lookup(sa_hdl, sa_attrs, cnt);
+
+	sa_handle_destroy(sa_hdl);
+	return (err);
+}
+
+int
+libuzfs_inode_setattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    const uzfs_attr_t *attr, uint64_t *txg)
+{
+	sa_handle_t *sa_hdl;
+	objset_t *os = dhp->os;
+	int err = sa_handle_get(os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	dmu_tx_t *tx = dmu_tx_create(os);
+	dmu_tx_hold_sa(tx, sa_hdl, B_FALSE);
+	if ((err = dmu_tx_assign(tx, TXG_WAIT)) != 0) {
+		dmu_tx_abort(tx);
+		return (err);
+	}
+
+	sa_bulk_attr_t sa_attrs[UZFS_END];
+	int cnt = 0;
+	libuzfs_add_bulk_attr(dhp, sa_attrs, &cnt, (uzfs_attr_t *)attr);
+	VERIFY0(sa_bulk_update(sa_hdl, sa_attrs, cnt, tx));
+
+	sa_handle_destroy(sa_hdl);
+
+	if (txg != NULL) {
+		*txg = tx->tx_txg;
+	}
+
+	dmu_tx_commit(tx);
+
+	return (0);
+}
+
+static int
+libuzfs_get_nvlist_from_handle(const sa_attr_type_t *sa_tbl,
+    nvlist_t **nvl, sa_handle_t *sa_hdl)
+{
+	int xattr_sa_size;
+	int err = sa_size(sa_hdl, sa_tbl[UZFS_XATTR], &xattr_sa_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	char *xattr_sa_data = vmem_alloc(xattr_sa_size, KM_SLEEP);
+	err = sa_lookup(sa_hdl, sa_tbl[UZFS_XATTR],
+	    xattr_sa_data, xattr_sa_size);
+	if (err == 0) {
+		err = nvlist_unpack(xattr_sa_data,
+		    xattr_sa_size, nvl, KM_SLEEP);
+	}
+	vmem_free(xattr_sa_data, xattr_sa_size);
+
+	return (err);
+}
+
+static int
+libuzfs_kvattr_set_nvlist(nvlist_t *nvl, const char *name,
+    const char *value, size_t value_size, size_t *xattr_sa_size)
+{
+	int err = nvlist_add_byte_array(nvl, name,
+	    (uchar_t *)value, (uint_t)value_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+	if (err != 0) {
+		return (err);
+	}
+
+	if (*xattr_sa_size > DXATTR_MAX_SA_SIZE ||
+	    *xattr_sa_size > SA_ATTR_MAX_LEN) {
+		err = nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
+		ASSERT(err != ENOENT);
+		if (err == 0) {
+			err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+		}
+		if (err != 0) {
+			return (err);
+		}
+		return (EFBIG);
+	}
+
+	return (0);
+}
+
+static int
+libufzs_inode_lookup_kvattr_zap(sa_handle_t *sa_hdl, sa_attr_type_t *sa_tbl,
+    const char *name, uint64_t *xattr_zap_obj)
+{
+	objset_t *os = sa_hdl->sa_os;
+	int err = sa_lookup(sa_hdl, sa_tbl[UZFS_ZXATTR],
+	    xattr_zap_obj, sizeof (*xattr_zap_obj));
+	if (err != ENOENT && err != 0) {
+		return (err);
+	}
+	if (err == 0) {
+		err = zap_contains(os, *xattr_zap_obj, name);
+	}
+	return (err);
+}
+
+static void
+libuzfs_inode_save_kvattr(sa_handle_t *sa_hdl, sa_attr_type_t *sa_tbl,
+    boolean_t existed_in_sa, boolean_t sa_space_enough, char *xattr_sa_data,
+    uint64_t xattr_sa_size, boolean_t existed_in_zap, uint64_t xattr_zap_obj,
+    uint64_t zap_entries_count, const char *name, const char *value,
+    size_t size, dmu_tx_t *tx)
```

Comment:
The parameters of this function looks horrible, compared with extracting such a function, I think moving the logic into libuzfs_inode_set_kvattr directly looks better.